### PR TITLE
added Wither API

### PR DESCRIPTION
**Pull request by Machine-Maker**
Various methods to affect Wither entities

### CODE REVIEW

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  */
+ public interface Wither extends Monster, Boss, RangedEntity { // Paper
++    // Paper start
++    /**
++     * Gets whether the wither is powered
++     * (aka cannot take ranged damage)
++     *
++     * @return powered state
++     */
++    boolean isPowered();
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+    boolean isCharged();
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+@@ -6,4 +6,42 @@ import com.destroystokyo.paper.entity.RangedEntity;
+  * Represents a Wither boss
+  */
+ public interface Wither extends Monster, Boss, RangedEntity { // Paper
++    // Paper start
++    /**
++     * Gets whether the wither is powered
++     * (aka cannot take ranged damage)
++     *
++     * @return powered state
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * @return whether the wither is charged
```

The Minecraft gamepedia wiki uses `shielded`. Choose charged or shielded and be consistent with it, I suppose.

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
++     */
++    boolean isPowered();
++
++    /**
++     * Gets the ticks the wither is invulnerable for
++     * This is the state at the beginning of fights
++     *
++     * @return invulnerable ticks
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * @return ticks the wither is invulnerable for
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
++     */
++    int getInvulnerableTicks();
++
++    /**
++     * Sets for how long in the future, the wither should be invulnerable
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * Sets for how long in the future, the wither should be invulnerable.
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
++     *
++     * @param ticks tick amount
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * @param ticks ticks the wither is invulnerable for
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
++     */
++    void setInvulnerableTicks(int ticks);
++
++    /**
++     * Gets whether or not this wither can travel through portals
++     *
++     * @return true if allowed, false if not
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * @return whether the wither can travel through portals
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
++     */
++    boolean canTravelThroughPortals();
++
++    /**
++     * Sets whether this wither can travel through portals
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * Sets whether the wither can travel through portals.
```

File: Spigot-API-Patches/0218-added-Wither-API.patch [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
++     *
++     * @param value true to allow travel, false to deny
```
Comment (Proximyst) [inline Spigot-API-Patches/0218-added-Wither-API.patch]:
```suggestion
+     * @param value whether the wither can travel through portals
```

File: Spigot-Server-Patches/0538-added-Wither-API.patch [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+     private static final PathfinderTargetCondition bG = (new PathfinderTargetCondition()).a(20.0D).a(EntityWither.bF);
++    private boolean canPortal = false; // Paper
+ 
+     public EntityWither(EntityTypes<? extends EntityWither> entitytypes, World world) {
+         super(entitytypes, world);
+@@ -529,6 +530,7 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
+         this.datawatcher.set((DataWatcherObject) EntityWither.bv.get(i), j);
+     }
+ 
++    public boolean isPowered() { return this.T_(); } // Paper - OBFHELPER
```
Comment (Proximyst) [inline Spigot-Server-Patches/0538-added-Wither-API.patch]:
```suggestion
+    public final boolean isPowered() { return this.T_(); } // Paper - OBFHELPER
```

File: Spigot-Server-Patches/0538-added-Wither-API.patch [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+         return this.getHealth() <= this.getMaxHealth() / 2.0F;
+     }
+@@ -545,8 +547,9 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
+ 
+     @Override
+     public boolean canPortal() {
+-        return false;
++        return canPortal; // Paper
+     }
++    public void setCanPortal(boolean canPortal) { this.canPortal = canPortal; } // Paper
```
Comment (Proximyst) [inline Spigot-Server-Patches/0538-added-Wither-API.patch]: Move up to field declaration.

File: Spigot-Server-Patches/0538-added-Wither-API.patch [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+     public boolean d(MobEffect mobeffect) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftWither.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftWither.java
+index 03cf6136e64bab7180933e106234b92c1255268d..35fcb9eaf53c31afbdd42a36a618285db8b272a7 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftWither.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftWither.java
+@@ -39,4 +39,30 @@ public class CraftWither extends CraftMonster implements Wither, CraftRangedEnti
+     public BossBar getBossBar() {
+         return bossBar;
+     }
++    // Paper start
```
Comment (Proximyst) [inline Spigot-Server-Patches/0538-added-Wither-API.patch]:
```suggestion
     }
+
+    // Paper start
```

**Review by Proximyst — CHANGES_REQUESTED** [review]

**Comment by Machine-Maker** [comment]
Ok, this was rebased.

**Review by Proximyst — APPROVED** [review]
LGTM! Sorry for the review taking a while 😝

**Pull request merged**